src-tauri/src/task_executor.rs

```diff
@@ -137,8 +139,27 @@ async fn execute_task(app_handle: &AppHandle, task: &mut Task) -> Result<Status>
     task.execution_chat_id = Some(chat.id);
     app_handle.emit_all("tasks:updated", &task)?;
 
+    let execution_steps_limit = {
+        let agent = repo::agents::get_for_chat(&*pool, chat.id).await?;
+        let settings_state: State<'_, RwLock<Settings>> = app_handle.state();
+        let settings = settings_state.read().await;
+
+        agent
+            .execution_steps_limit
+            .unwrap_or(settings.agents.execution_steps_limit)
+    };
+
     // TODO: refactor this loop
     loop {
+        let messages_count = repo::messages::get_count_of_failed_messages(&*pool, chat.id).await?;
+        if messages_count >= execution_steps_limit {
+            repo::tasks::fail(&*pool, task.id).await?;
+
+            task.status = Status::Failed;
+            app_handle.emit_all("tasks:updated", &task)?;
+
+            return Err(Error::LimitExpired.into());
+        }
```

Comment:
It's a caller responsibility in this case to update the task state. Just return an error and you're good.

```suggestion
        if messages_count >= execution_steps_limit {
            return Err(Error::StepsLimitExceeded(execution_steps_limit).into());
        }
```